runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -0,0 +1,222 @@
+local util = require('vim.lsp.util')
+local log = require('vim.lsp.log')
+local api = vim.api
+local M = {}
+
+-- [[
+-- TODO:
+-- [ ] only request for visible range
+-- ]]
+--
+---@class lsp.inlay_hint.bufstate
+---@field version integer
+---@field client_hint table<integer, table<integer, lsp.InlayHint[]>> client_id -> (lnum -> hints)
+
+---@type table<integer, lsp.inlay_hint.bufstate>
+local hint_cache_by_buf = setmetatable({}, {
+  __index = function(t, b)
+    local key = b > 0 and b or api.nvim_get_current_buf()
+    return rawget(t, key)
+  end,
+})
+
+--- Store hints for a specific buffer and client
+--- Resolves unresolved hints
+---
+---@param hints lsp.InlayHint[]|nil (table) of lenses to store
+---@param bufnr integer
+---@param client_id integer
+---@param version integer
+function M.save(hints, bufnr, client_id, version)
+  if not hints then
+    return
+  end
+  local bufstate = hint_cache_by_buf[bufnr]
+  if not bufstate then
+    bufstate = {
+      client_hint = vim.defaulttable(),
+      version = version,
+    }
+    hint_cache_by_buf[bufnr] = bufstate
+    api.nvim_buf_attach(bufnr, false, {
+      on_detach = function(_, b)
+        hint_cache_by_buf[b] = nil
+      end,
+      on_reload = function(_, b)
+        hint_cache_by_buf[b] = nil
+      end,
+    })
+  end
+  local hints_by_client = bufstate.client_hint
+  local hints_by_lnum = hints_by_client[client_id]
+  local client = vim.lsp.get_client_by_id(client_id)
+
+  -- To reduce flicker, preserve old resolved hints until new ones are resolved.
+
+  local new_hints_by_lnum = vim.defaulttable()
+  local num_unprocessed = #hints
+  if num_unprocessed == 0 then
+    hints_by_client[client_id] = {}
+    bufstate.version = version
+    return
+  end
+
+  for _, hint in ipairs(hints) do
+    local lnum = hint.position.line
+    table.insert(new_hints_by_lnum[lnum], hint)
```

Comment:
```suggestion
    new_hints_by_lnum[lnum][#new_hints_by_lnum[lnum] + 1] =  hint
```

Reply:
Kindly refrain from making such stylistic suggestions without proper basis for it.

1. If it makes a performance difference, show some (any) numbers with your suggestion.
2. If we care about the style, we should add it to a style guide and make it objective ("only uses table.insert for nested tables").
3. Otherwise, let the PR author do what they like.